### PR TITLE
[SPARK-14127][SQL] Native "DESC [EXTENDED | FORMATTED] <table>" DDL command

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

This PR implements native `DESC [EXTENDED | FORMATTED] <table>` DDL command. Sample output:

```
scala> spark.sql("desc extended src").show(100, truncate = false)
+----------------------------+---------------------------------+-------+
|col_name                    |data_type                        |comment|
+----------------------------+---------------------------------+-------+
|key                         |int                              |       |
|value                       |string                           |       |
|                            |                                 |       |
|# Detailed Table Information|CatalogTable(`default`.`src`, ...|       |
+----------------------------+---------------------------------+-------+


scala> spark.sql("desc formatted src").show(100, truncate = false)
+----------------------------+----------------------------------------------------------+-------+
|col_name                    |data_type                                                 |comment|
+----------------------------+----------------------------------------------------------+-------+
|key                         |int                                                       |       |
|value                       |string                                                    |       |
|                            |                                                          |       |
|# Detailed Table Information|                                                          |       |
|Database:                   |default                                                   |       |
|Owner:                      |lian                                                      |       |
|Create Time:                |Mon Jan 04 17:06:00 CST 2016                              |       |
|Last Access Time:           |Thu Jan 01 08:00:00 CST 1970                              |       |
|Location:                   |hdfs://localhost:9000/user/hive/warehouse_hive121/src     |       |
|Table Type:                 |MANAGED                                                   |       |
|Table Parameters:           |                                                          |       |
|  transient_lastDdlTime     |1451898360                                                |       |
|                            |                                                          |       |
|# Storage Information       |                                                          |       |
|SerDe Library:              |org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe        |       |
|InputFormat:                |org.apache.hadoop.mapred.TextInputFormat                  |       |
|OutputFormat:               |org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat|       |
|Num Buckets:                |-1                                                        |       |
|Bucket Columns:             |[]                                                        |       |
|Sort Columns:               |[]                                                        |       |
|Storage Desc Parameters:    |                                                          |       |
|  serialization.format      |1                                                         |       |
+----------------------------+----------------------------------------------------------+-------+
```
## How was this patch tested?

A test case is added to `HiveDDLSuite` to check command output.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -1012,7 +1017,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
       entry("field.delim", ctx.fieldsTerminatedBy) ++
         entry("serialization.format", ctx.fieldsTerminatedBy) ++
         entry("escape.delim", ctx.escapedBy) ++
-        entry("colelction.delim", ctx.collectionItemsTerminatedBy) ++
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: A typo bug, not related to this PR.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Oh wait... It's actually a typo from Hive... :dizzy_face:

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -48,6 +48,7 @@ case class CatalogStorageFormat(
     inputFormat: Option[String],
     outputFormat: Option[String],
     serde: Option[String],
+    compressed: Boolean,
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Is this ever true? If it isn't we could leave it out.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Nvm. Hive can pass compressed tables.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 3]
```diff
@@ -290,29 +291,88 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
...
+  }
+
+  private def describeExtended(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
+    describe(relation, buffer)
+
+    buffer += Row("", "", "")
+    buffer += Row("# Detailed Table Information", relation.catalogTable.toString, "")
+  }
+
+  private def describeFormatted(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: We could also use some sort of an `append(string, string, string)` function instead of calling `buffer += Row(..., ..., ...)` alot. But i guess that is just a matter of preference.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 4]
```diff
@@ -290,29 +291,88 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
...
+    buffer += Row("Storage Desc Parameters:", "", "")
+    table.storage.serdeProperties.foreach { case (key, value) =>
+      buffer += Row(s"  $key", value, "")
+    }
+  }
+
+  private def describeSchema(schema: StructType, buffer: ArrayBuffer[Row]): Unit = {
+    schema.foreach { column =>
+      val comment =
+        if (column.metadata.contains("comment")) column.metadata.getString("comment") else ""
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Just an idea: So 9/10 times I use `column.metadata` it is to get the comment. Shouldn't we just add this to the StructField...?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -1012,6 +1017,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
       entry("field.delim", ctx.fieldsTerminatedBy) ++
         entry("serialization.format", ctx.fieldsTerminatedBy) ++
         entry("escape.delim", ctx.escapedBy) ++
+        // The following typo is inherited from Hive...
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: +1 for this comment. Totally got me the last time.

**Comment by hvanhovell** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 3 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Makes sense, actually there were several times that I forgot to add the trailing empty string(s) while working on this PR. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 4 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: I think the metadata is heavily used in ML code. Another thing is that the data type API is already public for a long time. We probably don't want to change it unless there are particular good reasons.

**Comment by liancheng** [comment]
Thanks for the review! I'm merging this to master and branch-2.0.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: I am not saying that we should (re)move it. I am only suggesting that it might be easier if we had such an accessor in StructField.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 6]
```diff
@@ -290,29 +291,92 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
...
+      append(buffer, s"# ${output(0).name}", output(1).name, output(2).name)
+      describeSchema(relation.catalogTable.partitionColumns, buffer)
+    }
+  }
+
+  private def describeExtended(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
+    describe(relation, buffer)
+
+    append(buffer, "", "", "")
+    append(buffer, "# Detailed Table Information", relation.catalogTable.toString, "")
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: @liancheng To improve the output of `Explain`, I plan to change the default implementation of `toString` of `case class CatalogTable`. That will also affect the output of `Describe Extended`. 

I checked what Hive did for the command `Describe Extended`, as follows. 

`Detailed Table Information	Table(tableName:t1, dbName:default, owner:root, createTime:1462627092, lastAccessTime:0, retention:0, sd:StorageDescriptor(cols:[FieldSchema(name:col, type:int, comment:null)], location:hdfs://6b68a24121f4:9000/user/hive/warehouse/t1, inputFormat:org.apache.hadoop.mapred.TextInputFormat, outputFormat:org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat, compressed:false, numBuckets:-1, serdeInfo:SerDeInfo(name:null, serializationLib:org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, parameters:{serialization.format=1}), bucketCols:[], sortCols:[], parameters:{}, skewedInfo:SkewedInfo(skewedColNames:[], skewedColValues:[], skewedColValueLocationMaps:{}), storedAsSubDirectories:false), partitionKeys:[], parameters:{transient_lastDdlTime=1462627092}, viewOriginalText:null, viewExpandedText:null, tableType:MANAGED_TABLE)`

Basically, in the implementation of `toString`, I will try to follow what you did in `describeFormatted` but the contents will be in a single line. Feel free to let me know if you have any concern or suggestion. Thanks!
